### ISSUE TITLE
AMP should enforce MAX_KEY_LENGTH in the protocol parser

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3478]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#3478|
|Type|enhancement|
|Created|2008-10-16 19:20:10Z|
|Branch|https://github.com/twisted/twisted/tree/amp-key-enforcer-3478|

In implementing AMP in other languages, I've discovered a common sort of error is getting the length calculation wrong and writing too many or too few bytes.  It's often difficult to figure out where the protocol went wrong, because the failure mode of the python implementation in the face of much bad input is to just leave the connection open forever, assuming that it is waiting for 2573 bytes in a key if, for example, a value tacks on a garbage "\r\n".

The response to a key that's too long should be immediately blowing up the connection as soon as the non-zero first length byte is received.  It's like the trailing comma in netstrings: that one little extra bit of redundancy can be really helpful.

<details><summary>Searchable metadata</summary>

```
trac-id__3478 3478
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_amp_key_enforcer_3478 branches/amp-key-enforcer-3478
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1224184810000000 1224184810000000
changetime__1225595473000000 1225595473000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [25237]) Branching to 'amp-key-enforcer-3478'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

That should do.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph set owner to @exarkun|
|-|-|

  1. `lengthLimitedExceeded` is a really weird name.  Why not `lengthLimitExceeded`?  Given that this is the name used in the test, it almost seems like this is a typo.
  1. It strikes me that rather than making the point of hookability `lengthLimitExceeded`, it should be `lengthReceived`.  Is there some reason not to do it that way?

I hope the AMP tests can remain unchanged with my suggested change...

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

Looks like the stringreceiver tests need to change anyway:
```
[FAIL]: twisted.test.test_stateful.Int32TestCase.test_lengthLimitExceeded

Traceback (most recent call last):
  File "/home/glyph/Projects/Twisted/trunk/twisted/test/test_protocols.py", line 434, in test_lengthLimitExceeded
    self.assertEqual(length, [11])
twisted.trial.unittest.FailTest: not equal:
a = []
b = [11]
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [25246]) rename `lengthLimitedExceeded` to `lengthLimitExceeded`

refs [#3478](https://github.com/twisted/twisted/issues/3478)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [25247]) make the other int32stringreceiver implementation support lineLengthExceeded

refs [#3478](https://github.com/twisted/twisted/issues/3478)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun removed owner|
|-|-|

  1. Yea, that was some kind of typo.  Renamed `lengthLimitExceeded`.
  2. I think the hook could be generalized to `lengthReceived` later without breaking compatibility (base implementation would enforce the limit and call `lengthLimitExceeded`, override for something else if you want).  I didn't generalize it here because AMP doesn't need to know about every length prefix received and it would be necessary to add control flow support to the `lengthReceived` hook in order to implement AMP's desired policy.   Perhaps it can be argued that `lengthLimitExceeded` should altering the control flow too, though...

I fixed `test_stateful` by implementing the same hook in its private re-implementation of `Int32StringReceiver`.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph set owner to @exarkun|
|-|-|

OK.  Looking good.  Merge.  Maybe make a (low-priority) ticket for the `lengthReceived` thing?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [25267]) Merge amp-key-enforcer-3478

Author: exarkun
Reviewer: glyph
Fixes: [#3478](https://github.com/twisted/twisted/issues/3478)

Change `BinaryBoxProtocol` so that as soon as it receives a key length
prefix larger than 255 it closes the connection and reports a protocol
violation as the reason for any failures.  Keys are restricted to 255
or fewer bytes by the protocol; early disconnection for violation of this
simplifies tracking down bugs in protocol implementations.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
